### PR TITLE
DO NOT MERGE YET: Rdar 30961871 metrics

### PR DESCRIPTION
**Pull request by graydon**
This set of changes augments ParseableOutput with a bit of cheap metrics-reporting machinery,
eventually destined for some form of telemetry Xcode-side (work there is still TBD, so don't merge
yet!)

The metrics come in two forms: a general rusage (user/sys/rss) dictionary reported for each
subprocess the driver runs, and a final set of counters representing jobs run and dependency
types that caused them (along with a cumulative rusage for the whole child process tree).

Feedback welcome. I expect this isn't _entirely_ right but I tried a few other arrangements and this
seems like the least-intrusive way of getting a small set of lightweight cost signals out of compilers
in the field.

rdar://30961871

### CODE REVIEW

**Comment by ddunbar** [comment]
This looks awesome to me, I think it has all the things I want!

Can we get an update to:
  https://github.com/apple/swift/blob/master/docs/DriverParseableOutput.rst
showing an example of the format and a description of what the key fields mean as well?

File: include/swift/Basic/TaskQueue.h [hunk 1]
```diff
@@ -47,6 +48,13 @@ class TaskQueue {
   unsigned NumberOfParallelTasks;
 
 public:
+
+  // \brief Get the resources used by just the current process.
+  static llvm::Optional<ResourceStats> OwnResourceStats();
+
+  // \brief Get the cumulative resources used by all children.
+  static llvm::Optional<ResourceStats> ChildResourceStats();
```
Comment (jrose-apple) [inline include/swift/Basic/TaskQueue.h]: Nitpick: Method names should have lowerCamelCase.

File: lib/Basic/Default/TaskQueue.inc [hunk 2]
```diff
@@ -52,6 +52,18 @@ public:
...
+TaskQueue::OwnResourceStats()
+{
+  return Optional<ResourceStats>();
+}
+
+Optional<ResourceStats>
+TaskQueue::ChildResourceStats()
+{
+  return Optional<ResourceStats>();
+}
```
Comment (jrose-apple) [inline lib/Basic/Default/TaskQueue.inc]: You forgot to update the Finished callback in the default TaskQueue implementation.

File: include/swift/Basic/TaskQueue.h [hunk 3]
```diff
@@ -98,12 +109,15 @@ class TaskQueue {
...
+  /// \param Resources an optional indication of resources consumed by the
+  /// subprocess during its execution
   ///
   /// \returns a TaskFinishedResponse indicating whether or not execution
   /// should proceed
   typedef std::function<TaskFinishedResponse(ProcessId Pid, StringRef ErrorMsg,
                                              StringRef Output, StringRef Errors,
-                                             void *Context, Optional<int> Signal)>
+                                             void *Context, Optional<int> Signal,
+                                             llvm::Optional<ResourceStats> Resources)>
```
Comment (jrose-apple) [inline include/swift/Basic/TaskQueue.h]: Nitpick: No "llvm::" needed.

File: lib/Driver/ParseableOutput.cpp [hunk 4]
```diff
@@ -73,6 +75,16 @@ namespace json {
       return seq[index];
     }
   };
+
+  template<>
+  struct ScalarTraits<llvm::sys::TimePoint<>> {
+    static void output(const llvm::sys::TimePoint<> &value, llvm::raw_ostream &os) {
+      using namespace std::chrono;
+      os << duration_cast<microseconds>(value.time_since_epoch()).count();
```
Comment (jrose-apple) [inline lib/Driver/ParseableOutput.cpp]: `time_since_epoch` isn't necessarily the same on all systems. We should probably pick our own epoch.

File: include/swift/Driver/DependencyGraph.h [hunk 5]
```diff
@@ -168,13 +175,20 @@ class DependencyGraphImpl {
...
   }
 
+  void
+  countMarking(CompilationCounters &counters,
+               DependencyMaskTy kinds, bool isCascading);
+
   void markTransitive(SmallVectorImpl<const void *> &visited,
-                      const void *node, MarkTracerImpl *tracer = nullptr);
+                      const void *node,
+                      CompilationCounters &counters,
```
Comment (jrose-apple) [inline include/swift/Driver/DependencyGraph.h]: Can you put this inside MarkTracer instead? Right now DependencyGraphImpl is still pretty generic and I like that we'd be able to break it out for some other purpose if we had to.

File: lib/Driver/DependencyGraph.cpp [hunk 6]
```diff
@@ -302,13 +304,43 @@ void DependencyGraphImpl::markExternal(SmallVectorImpl<const void *> &visited,
...
     visited.push_back(dependent.node);
-    markTransitive(visited, dependent.node);
+    markTransitive(visited, dependent.node, counters);
+  }
+}
+
+void
+DependencyGraphImpl::countMarking(CompilationCounters &counters,
+                                  DependencyMaskTy kinds, bool isCascading) {
+  if (isCascading) {
```
Comment (jrose-apple) [inline lib/Driver/DependencyGraph.cpp]: This data is measuring "dependency edges actually traversed during a rebuild", right? Is that the right thing to be measuring? (As opposed to "total kinds of dependency edges", or "max fan-out from one changed file, regardless of edges taken to get there".)

**Comment by graydon** [comment]
Closing this out in favour of #8477

**Pull request closed without merging**